Subject: [PATCH] IJPL-195388 Island Dark scheme is visible in the scheme list despite Islands theme are hidden
IJPL-195149 Hide the island theme under the registry key
IJPL-195149 Hide the island theme under the registry key
---
Index: platform/platform-impl/src/com/intellij/ide/ui/laf/ThemeListProviderImpl.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/ide/ui/laf/ThemeListProviderImpl.kt b/platform/platform-impl/src/com/intellij/ide/ui/laf/ThemeListProviderImpl.kt
--- a/platform/platform-impl/src/com/intellij/ide/ui/laf/ThemeListProviderImpl.kt	(revision a186cf25b150df78d501b819a6bf80edf0840ba8)
+++ b/platform/platform-impl/src/com/intellij/ide/ui/laf/ThemeListProviderImpl.kt	(revision 25455f00ae3682706da946b8a5bf62813dd9a766)
@@ -47,7 +47,7 @@
         else customThemes.add(info)
       }
     }
-    else if (ExperimentalUI.isNewUI()) {
+    else if (ExperimentalUI.isNewUI() && Registry.`is`("idea.islands.enabled", false)) {
       uiThemeProviderListManager.getThemeListForTargetUI(TargetUIType.ISLAND).forEach { info ->
         if (!info.isThemeFromPlugin) islandUiThemes.add(info)
         else customThemes.add(info)
Index: platform/platform-impl/src/com/intellij/openapi/application/impl/islands/IslandsUICustomization.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/openapi/application/impl/islands/IslandsUICustomization.kt b/platform/platform-impl/src/com/intellij/openapi/application/impl/islands/IslandsUICustomization.kt
--- a/platform/platform-impl/src/com/intellij/openapi/application/impl/islands/IslandsUICustomization.kt	(revision a186cf25b150df78d501b819a6bf80edf0840ba8)
+++ b/platform/platform-impl/src/com/intellij/openapi/application/impl/islands/IslandsUICustomization.kt	(revision 32b15b936175505feda00571d6aecb2bf584951f)
@@ -1,12 +1,17 @@
 // Copyright 2000-2025 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
 package com.intellij.openapi.application.impl.islands

+import com.intellij.diagnostic.LoadingState
+import com.intellij.ide.AppLifecycleListener
 import com.intellij.ide.impl.ProjectUtil
+import com.intellij.ide.ui.LafManager
 import com.intellij.ide.ui.LafManagerListener
+import com.intellij.ide.util.PropertiesComponent
 import com.intellij.openapi.actionSystem.ex.ActionButtonLook
 import com.intellij.openapi.application.ApplicationManager
 import com.intellij.openapi.application.impl.InternalUICustomization
 import com.intellij.openapi.application.impl.ToolWindowUIDecorator
+import com.intellij.openapi.editor.colors.EditorColorsManager
 import com.intellij.openapi.fileEditor.FileEditorManager
 import com.intellij.openapi.fileEditor.impl.EditorEmptyTextPainter
 import com.intellij.openapi.fileEditor.impl.EditorsSplitters
@@ -14,6 +19,8 @@
 import com.intellij.openapi.ui.OnePixelDivider
 import com.intellij.openapi.ui.Splittable
 import com.intellij.openapi.util.registry.Registry
+import com.intellij.openapi.util.registry.RegistryValue
+import com.intellij.openapi.util.registry.RegistryValueListener
 import com.intellij.openapi.wm.IdeFrame
 import com.intellij.openapi.wm.IdeGlassPane
 import com.intellij.openapi.wm.IdeGlassPaneUtil
@@ -108,6 +115,62 @@
         toolkit.addAWTEventListener(awtListener, AWTEvent.HIERARCHY_EVENT_MASK)
       }
     })
+
+    if (LoadingState.COMPONENTS_LOADED.isOccurred) {
+      checkThemesVisible()
+    }
+    else {
+      connection.subscribe(AppLifecycleListener.TOPIC, object : AppLifecycleListener {
+        override fun appStarted() {
+          checkThemesVisible()
+        }
+      })
+    }
+  }
+
+  private fun checkThemesVisible() {
+    val key = "idea.islands.enabled"
+    val properties = PropertiesComponent.getInstance()
+
+    if (!properties.getBoolean(key, false)) {
+      properties.setValue(key, true)
+
+      if (isIslandsEnabled) {
+        Registry.get(key).setValue(true)
+      }
+      else {
+        resetColorScheme()
+      }
+    }
+
+    Registry.get(key).addListener(object : RegistryValueListener {
+      override fun afterValueChanged(value: RegistryValue) {
+        if (!value.asBoolean()) {
+          if (isIslandsEnabled) {
+            val lafManager = LafManager.getInstance()
+            val theme = if (JBColor.isBright()) lafManager.defaultLightLaf else lafManager.defaultDarkLaf
+
+            if (theme != null) {
+              lafManager.setCurrentLookAndFeel(theme, true)
+
+              val colorsManager = EditorColorsManager.getInstance()
+              theme.installEditorScheme(colorsManager.getScheme(if (JBColor.isBright()) "Light" else "Dark") ?: colorsManager.defaultScheme)
+            }
+          }
+          else {
+            resetColorScheme()
+          }
+        }
+      }
+    }, ApplicationManager.getApplication())
+  }
+
+  private fun resetColorScheme() {
+    val colorsManager = EditorColorsManager.getInstance()
+
+    if (colorsManager.globalScheme.displayName == "Island Dark") {
+      colorsManager.setGlobalScheme(colorsManager.getScheme("Dark"))
+    }
   }

   private val tabPainterAdapter = ManyIslandsTabPainterAdapter()
Index: platform/util/resources/misc/registry.properties
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>ISO-8859-1
===================================================================
diff --git a/platform/util/resources/misc/registry.properties b/platform/util/resources/misc/registry.properties
--- a/platform/util/resources/misc/registry.properties	(revision a186cf25b150df78d501b819a6bf80edf0840ba8)
+++ b/platform/util/resources/misc/registry.properties	(revision 25455f00ae3682706da946b8a5bf62813dd9a766)
@@ -2564,6 +2564,10 @@
 station.use.station.comms.tools.management=false
 station.use.station.comms.tools.management.description=Use station backend for updating/installing tools

+idea.islands.enabled=false
+idea.islands.enabled.description=Enable Islands UI themes
+idea.islands.enabled.restartRequired=true
+
 idea.islands.gradient.enabled=true
 idea.islands.gradient.enabled.description=Enable Islands UI themes gradient painter
 idea.islands.gradient.enabled.restartRequired=true
Index: platform/platform-impl/src/com/intellij/openapi/editor/colors/impl/EditorColorSchemesSorterImpl.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/platform/platform-impl/src/com/intellij/openapi/editor/colors/impl/EditorColorSchemesSorterImpl.kt b/platform/platform-impl/src/com/intellij/openapi/editor/colors/impl/EditorColorSchemesSorterImpl.kt
--- a/platform/platform-impl/src/com/intellij/openapi/editor/colors/impl/EditorColorSchemesSorterImpl.kt	(revision 30463a7bc4ee353518b7aba77cadc6687aa2356f)
+++ b/platform/platform-impl/src/com/intellij/openapi/editor/colors/impl/EditorColorSchemesSorterImpl.kt	(revision 32b15b936175505feda00571d6aecb2bf584951f)
@@ -13,6 +13,7 @@
 import com.intellij.openapi.editor.colors.EditorColorsScheme
 import com.intellij.openapi.editor.colors.Groups
 import com.intellij.openapi.options.Scheme
+import com.intellij.openapi.util.registry.Registry
 import com.intellij.ui.ExperimentalUI
 import org.jetbrains.annotations.ApiStatus

@@ -78,6 +79,9 @@
       }
       schemesToFilterOut.addAll(newUiSchemeIds)
     }
+    else if (!Registry.`is`("idea.islands.enabled", false)) {
+      schemesToFilterOut.add("Island Dark")
+    }

     schemesToFilterOut.forEach {
       schemes.remove(it)
